domain/application/service/storage.go

```diff
@@ -0,0 +1,107 @@
+	// - [applicationerrors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+	// - [applicationerrors.InvalidStorageCount]: when the allowed attachment count would be violated.
+	// - [applicationerrors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+	AttachStorage(ctx context.Context, storageID storage.ID, unitUUID unit.UUID) error
+
+	// AddStorageForUnit attaches the specified storage to the specified unit.
+	// The following error types can be expected:
+	// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+	// - [applicationerrors.UnitNotAlive]: when the unit is not alive.
+	// - [applicationerrors.StorageNotAlive]: when the storage is not alive.
+	// - [applicationerrors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+	// - [applicationerrors.InvalidStorageCount]: when the allowed attachment count would be violated.
+	// - [applicationerrors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+	AddStorageForUnit(
+		ctx context.Context, storageName storage.Name, unitUUID unit.UUID, stor storage.Directive,
+	) ([]storage.ID, error)
+
+	// DetachStorageForUnit detaches the specified storage from the specified unit.
+	// The following error types can be expected:
+	// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+	// - [applicationerrors.StorageNotDetachable]: when the type of storage is not detachable.
+	DetachStorageForUnit(ctx context.Context, storageID storage.ID, unitUUID unit.UUID) error
+
+	// DetachStorage detaches the specified storage from whatever node it is attached to.
+	// The following error types can be expected:
+	// - [applicationerrors.StorageNotDetachable]: when the type of storage is not detachable.
+	DetachStorage(ctx context.Context, storageID storage.ID) error
+}
+
+// AttachStorage attached the specified storage to the specified unit.
+// The following error types can be expected:
+// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+// - [applicationerrors.StorageAttachmentAlreadyExistsError]: when the attachment already exists.
+// - [applicationerrors.WrongStorageOwnerError]: when the storage owner does not match the unit.
+// - [applicationerrors.UnitNotAlive]: when the unit is not alive.
+// - [applicationerrors.StorageNotAlive]: when the storage is not alive.
+// - [applicationerrors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [applicationerrors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [applicationerrors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AttachStorage(ctx context.Context, storageID storage.ID, unitName unit.Name) error {
```

Comment:
We should validate the ID and the unit name, before attempting a request on state.